circuits/cpp/barretenberg/cpp/src/barretenberg/stdlib/primitives/biggroup/biggroup_goblin.hpp

```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+namespace proof_system::plonk {
+namespace stdlib {
+
+/**
+ * @brief Goblin style batch multiplication
+ *
+ * @details In goblin-style arithmetization, the operands (points/scalars) for each mul-accumulate operation are
+ * decomposed into smaller components and written to an operation queue via the builder. The components are also added
+ * as witness variables. This function adds constraints demonstrating the fidelity of the point/scalar decompositions
+ * given the indices of the components in the variables array. The actual mul-accumulate operations are performed
+ * natively (without constraints) under the hood, and the final result is obtained by queueing an equality operation via
+ * the builder. The components of the result are returned as indices into the variables array from which the resulting
+ * accumulator point is re-constructed.
+ *
+ * @tparam C CircuitBuilder
+ * @tparam Fq Base field
+ * @tparam Fr Scalar field
+ * @tparam G Native group
+ * @param points
+ * @param scalars
+ * @param max_num_bits
+ * @return element<C, Fq, Fr, G>
+ */
+template <typename C, class Fq, class Fr, class G>
+element<C, Fq, Fr, G> element<C, Fq, Fr, G>::goblin_batch_mul(const std::vector<element>& points,
+                                                              const std::vector<Fr>& scalars,
+                                                              [[maybe_unused]] const size_t max_num_bits)
+{
+    auto builder = points[0].get_context();
+
+    // Check that the internal accumulator is zero?
+    ASSERT(builder->op_queue.get_accumulator().is_point_at_infinity());
+
+    // Loop over all points and scalars
+    size_t num_points = points.size();
+    for (size_t i = 0; i < num_points; ++i) {
+        auto& point = points[i];
+        auto& scalar = scalars[i];
+
+        // Populate the goblin-style ecc op gates for the given mul inputs
+        auto op_tuple = builder->queue_ecc_mul_accum(point.get_value(), scalar.get_value());
+
+        // Adds constraints demonstrating proper decomposition of point coordinates.
+        // Note: may need to do point.x.assert_is_in_field() prior to the assert_eq() according to Kesha.
```

Comment:
The mental model I have for assert equal is "this is a memcpy but for circuits", in particular:

- memcpy does not assert things are equal
- and memcpy(a,b) is not the same as memcpy(b,a)

Reply:
hmm yeah taking a closer look there is a big DO NOT USE IN CIRCUITS warning. Definitely don't understand the details here but it's on the list of things to figure out. Either way the fact that this mistake can be made is a bad sign